Replace Font Awesome icons in Journalist Interface with PNGs

DrGFreeman opened the pull request:
## Status

Ready for review.

## Description of Changes

Fixes #5546
Fixes #5576

Changes proposed in this pull request:

**Replace Font Awesome icons with PNGs in Journalist Interface.**
    
Replace all Font Awesome icons in JI with PNGs so they are displayed correctly with Tor Browser in "Safest" setting (#5546 & #5576).  See screenshots in comment below.
    
**Related changes included:**
- Increase font size of source submission counts text in `/`
- Ensure consistent vertical spacing in users table in `/admin/`

**Other changes:**
- Capitalize "F" in "Filter by codename" placeholder text in `/` (#5576)
- Add "«" in "Back to admin interface" link text in `/admin/edit/<user_id>` for consistency with other admin pages.

## Testing

How should the reviewer test this PR?
Write out any special testing steps here.

**Journalist homepage `/`**

Login to Journalist Interface using Tor Browser with security settings set to "Safest".
- [ ] Verify "Filter by codename" placeholder text of input box has capital "F".
- [ ] Verify all icons display and are properly positioned.
- [ ] Hover on Delete button and verify white trash icon is visible on hover.
- [ ] Hover on source star and verify orange star is visible.
- [ ] Click star on a source and verify the star icon turns to orange.
- [ ] Verify nothing is abnormal with the page layout.

**Source submissions page `/col/<filesystem_id>`**

Login source interface in a new browser tab.
Submit one message.
Submit one file.

Refresh the JI browser tab.
Click on the new source's codename.
- [ ] Verify the chevron icon displays between the All Source link and the source codename.
- [ ] Verify all button icons display and are properly positioned.
- [ ] Hover on Delete Selected button and verify the white trash icon is visible.
- [ ] Verify nothing is abnormal with the page layout.
Download one source message.
Submit two replies.

Refresh Source Interface browser tab.
Delete one reply.

Refresh the JI browser tab.
- [ ] Verify all icons display left of the source codename (open envelope, closed envelope, reply (unread), check (read reply) and are properly positioned.
- [ ] Verify the rightmost icons display (files, messages, reply) and are properly positioned.
- [ ] Scroll to bottom and verify trash icon displays on DELETE SOURCE AND SUBMISSIONS button.
- [ ] Hover on DELETE SOURCE AND SUBMISSIONS button and verify that white trash icon is displayed.

**Admin page `/admin/`**

Click the Admin link.
- [ ] Verify the plus and pencil icons display in the ADD USER and INSTANCE CONFIG buttons.
- [ ] Verify the edit and trash icons display on each journalist row.
- [ ] Verify the vertical spacing between the rows is similar.
- [ ] Hover on the second journalist trash icon and verify the darker red trash icon displays.
- [ ] Verify nothing is abnormal with the page layout

**Add User page `/admin/add`**

Click the ADD USER button.
- [ ] Verify the plus icon displays in the ADD USER button.
- [ ] Verify nothing is abnormal with the page layout.
Click the *Back to admin interface* link.
Additional
**Edit User page `/admin/edit/<journalist_id>`**

Click the edit icon on one journalist row.
- [ ] Verify the "«" character displays at the begining of the *Back to admin interface* link.
- [ ] Verify the "sync" icon displays on the RESET PASSWORD, RESET MOBILE APP CREDENTIALS and RESET SECURITY KEY CREDENTIALS buttons.
- [ ] Verify nothing is abnormal with the page layout.
Click the *Back to admin interface* link.

**Instance Configuration page `/admin/config`**

Click the INSTANCE CONFIG button.
- [ ] Verify the bell icon displays on the SEND TEST OSSEC ALERT button.
- [ ] Verify the edit (pencil) icon displays on the UPDATE LOGO and UPDATE SUBMISSION PREFERENCES buttons.
- [ ] Verify nothing is abnormal with the page layout.

**Flag source for reply success page `/flag`**

Shutdown the server and temporarily change line 250 of `securedrop/source_app/main.py` from `if entropy_avail >= 2400:` to `if entropy_avail >= 2400 and False:`.
Restart the server, log in as a new source and submit a message.
View this new source via the Journalist Interface.
Click FLAG THIS SOURCE FOR REPLY.
- [ ] Verify the info circle icon is displays and is properly positioned.
- [ ] Verify nothing is abnormal with the page layout.

## Deployment

Any special considerations for deployment? Consider both:

1. Upgrading existing production instances.
2. New installs.

## Checklist

### If you made changes to the server application code:

- [x] Linting (`make lint`) and tests (`make test`) pass in the development container

### If you made changes to `securedrop-admin`:

- [ ] Linting and tests (`make -C admin test`) pass in the admin development container

### If you made changes to the system configuration:

- [ ] [Configuration tests](https://docs.securedrop.org/en/latest/development/testing_configuration_tests.html) pass

### If you made non-trivial code changes:

- [x] I have written a test plan and validated it for this PR

### If you made changes to documentation:

- [ ] Doc linting (`make docs-lint`) passed locally

### If you added or updated a code dependency:

Choose one of the following:

- [ ] I have performed a diff review and pasted the contents to [the packaging wiki](https://github.com/freedomofpress/securedrop-debian-packaging/wiki)
- [ ] I would like someone else to do the diff review

DrGFreeman committed:
Capitalize F in JI "Filter by codename" placeholder text

Ref. #5576.

DrGFreeman committed:
Replace Font Awesome icons with PNGs in JI

Replace all Font Awesome icons in JI with PNGs so they are displayed
correctly with Tor Browser in "Safest" setting (#5546 & #5576).

Related changes included:
- Increase font size of source submisison counts text in /
- Ensure consistent vertical spacing in users table in /admin/

DrGFreeman commented:
Screenshots taken with Tor Browser at "Safest" setting below:

**JI homepage** (Updated)

![image](https://user-images.githubusercontent.com/22901938/96399633-d3becf80-119c-11eb-9a17-b69de80cdad6.png)


**Source page**

![image](https://user-images.githubusercontent.com/22901938/96377583-7e5ed000-1154-11eb-88c0-a28753a22fd4.png)

**Admin interface**

![image](https://user-images.githubusercontent.com/22901938/96377606-b5cd7c80-1154-11eb-9c9c-dc555bd1b70d.png)

**Add user page**

![image](https://user-images.githubusercontent.com/22901938/96377640-e44b5780-1154-11eb-95e6-f46296b8f91f.png)

**Edit user page** (lower portion where icons are used)

![image](https://user-images.githubusercontent.com/22901938/96377694-3f7d4a00-1155-11eb-9894-87515d885f84.png)

**Config page**

![image](https://user-images.githubusercontent.com/22901938/96377724-83704f00-1155-11eb-8b21-231d859d6987.png)

**Flag source for reply success page**

![image](https://user-images.githubusercontent.com/22901938/96671288-70b17200-132f-11eb-9f3f-d16aa341bfff.png)

DrGFreeman commented:
## Notes

1. (Update: resolved in https://github.com/freedomofpress/securedrop/pull/5593/commits/161aebafe08c137eb3425aea51a3757fe61e6e7b) ~~Hover behavior on the source star in `/`, trash buttons and trash icon in `/admin/` is functional.  However, **I had to disable the tooltip on the trash icon in `/admin` to make the hover work**.  There is certainly a way to get it to work but this is beyond my knowledge of CSS at this point.~~
~~I suspect I would need to modify these [two selectors](https://github.com/DrGFreeman/securedrop/blob/ea36d78b56b0a2cd474a626f26183a0407a1edb0/securedrop/sass/modules/_button.sass#L118-L122) to also capture the images that are [inside the `<i>` element](https://github.com/DrGFreeman/securedrop/blob/ea36d78b56b0a2cd474a626f26183a0407a1edb0/securedrop/journalist_templates/admin.html#L43-L46) which is currently deactivated. I will read a bit on css selectors and try to make it work.~~

1. I have not cleaned up the repository of unused font-awesome files. I could include this in the changes if desired.

1. The `securedrop/static/i/` folder also contains many icons in PNG format I suspect are note used anymore (though some are used in flash messages). I could remove the unused icons if desired.

1. I put the new icons in `securedrop/static/icons` for now to avoid mixing up with the other icons in `securedrop/static/i` as per the previous note. I can move them inside `securedrop/static/i` if I clean the unused icons and ensure there is no filename collision.

1. (Update: addressed in https://github.com/freedomofpress/securedrop/pull/5593/commits/91b59e892c3bb593f785bcfe2a8fc5dc04c67c41) ~~The `securedrop/journalist_templates/flag.html` template uses the `fa-info-circle` icon. I have not replaced it yet as I could not figure how to access this page through the JI.~~

DrGFreeman commented:
@eloquence, this is a new PR replacing #5591 where I unfortunately messed-up the git history trying to rebase the branch. This one has a cleaner commit history.

I also added the new icon files to the AppArmor profile as [you requested](https://github.com/freedomofpress/securedrop/pull/5591#issuecomment-712465560).

I added a test plan in the PR as well.

I will will keep on the tooltip mentioned in my note above and then switch the PR to "Ready for Review".

Also, please let me know your opinion about the other points in my notes above (i.e. cleanup of unused icons & fonts).

eloquence commented:
> The securedrop/journalist_templates/flag.html template uses the fa-info-circle icon. I have not replaced it yet as I could not figure how to access this page through the JI.

Temporarily change line 250 `securedrop/source_app/main.py` from `if entropy_avail >= 2400:` to `if entropy_avail >= 2400 and False:`. Log in as a new source, and view this source via the Journalist Interface. Flag the source for reply, and you will see the confirmation message with the FA icon.

As you may have already noticed during development, until #5595 lands (and your PR is rebased onto it), hot reloading of the Journalist Interface does not work on current `develop`. That means that after making the change above, you have to restart the development environment for it to take effect.

eloquence commented:
This looks overall great already, and the new icons look super slick (thanks @ninavizz for recommending these replacements) and work well at high zoom levels.

One issue I'm noticing in testing is that due to the lack of a `width` attribute on many of the PNG uses, you'll see layout reflows when the page is loaded or reloaded:

![reload-issue](https://user-images.githubusercontent.com/213636/96655279-edbdf680-12f1-11eb-9e20-f85cfb0bf1d1.gif)

Note how everything jumps back and forth. You can prevent this by specifying both width and height consistently.

DrGFreeman committed:
Restore hover tooltip on journalist delete button

Change child combinator selector to descendant combinator to capture
<img> elements inside <i> element on button hover.

DrGFreeman commented:
Thanks for the feedback @eloquence! I still have a lot to learn and this kind of feedback is very useful! I had put only one dimension on the icons since they did not scale to exact integer values on both dimensions. I'll add the missing dimensions (width mostly), I presume the sight distortion from the rounding will not be noticeable.

I restored the hover "tooltip" on the delete journalist icon in `/admin` in https://github.com/freedomofpress/securedrop/pull/5593/commits/161aebafe08c137eb3425aea51a3757fe61e6e7b.

I'll also replace the `fa-info-circle` icon, checking it with the method you provided.

Let me know if you also want me to include cleanup of unused icons and font awesome files as part of this PR.

Thanks again!

DrGFreeman commented:
> As you may have already noticed during development, until #5595 lands (and your PR is rebased onto it), hot reloading of the Journalist Interface does not work on current `develop`. That means that after making the change above, you have to restart the development environment for it to take effect.

I had noticed. I got pretty good at pressing `Ctrl-C`, `Up Arrow`, `Enter` to restart the server then log back in the JI while working on this PR. I setup my KeePassXC to auto-fill the journalist credentials, including TOTP token to login quickly ;)

DrGFreeman committed:
Add missing dimensions to icon <img> elements

Add missing dimensions to icons <img> elements to avoid reflow on page
load / reload.

DrGFreeman committed:
Replace fa-info-circle icon with PNG in JI /flag

DrGFreeman commented:
Last JI icon replaced: `fa-info-circle` in `/flag`. Screenshot added with others in first comment of PR.

Corresponding test steps added to PR as well.

@eloquence FYI, switching to "Ready for Review".

DrGFreeman committed:
Add icon filename missing in AppArmor profile

The info-circle-blue-grey.png icon added in 91b59e892c3bb593f785bcfe2a8fc5dc04c67c41
is missing from the AppArmor profile. Add this icon to the file.

rmol commented on securedrop/journalist_templates/_source_row.html:
```diff
@@ -4,27 +4,39 @@
-      <button class="button-star starred" type="submit"
-              formaction="{{ url_for('col.remove_star', filesystem_id=source.filesystem_id) }}"><i class="fa fa-star"></i></button>
+      <button class="button-star starred" type="submit" formaction="{{ url_for('col.remove_star', filesystem_id=source.filesystem_id) }}">
+        <img src="{{ url_for('static', filename='icons/starred.png') }}" class="icon-drop icon-star-source" width="16" height="15" alt="starred-icon">
```
The `alt` attributes are a great addition, but we should translate them, so they're accessible in other languages. For example, `alt="{{ gettext("starred icon") }}"`.

DrGFreeman replied:
`gettext` added for all non-blank `alt` attributes.

rmol commented on securedrop/journalist_templates/_source_row.html:
```diff
@@ -4,27 +4,39 @@
+      {{ ngettext('1 doc', '{num} docs', docs).format(num=docs) }}
+    </span>
+    <span>
+      <img src="{{ url_for('static', filename='icons/messages.png') }}" class="icon-drop" width="14" height="13" alt="messages-icon">
```
Just writing this is causing me some self-loathing, but the messages icon is enough shorter than the files icon that it caught my eye. Compare it with a height of 15. To me they're better aligned, but the stretching does increase the weight of the lines in the messages icon. If no one else sees a need to change it I will happily agree. :man_shrugging: 

Height 13:

![source-list-messages-icon-height-13](https://user-images.githubusercontent.com/47430598/96788331-5627d900-13c1-11eb-92fa-75bd0a827942.png)

Height 15:

![source-list-messages-icon-height-15](https://user-images.githubusercontent.com/47430598/96788340-5aec8d00-13c1-11eb-9a9a-178d19d483b7.png)

eloquence replied:
My own bias is always to respect an icon's ratio if possible, but I'm fine either way. Flagging to @ninavizz for visibility.

DrGFreeman replied:
I had initially put the messages icon the same height as the files icon (16px, natural width of 17.9px) but found it appear much larger than the files icon (14 x 16px). I reduced it, probably too much.

Since this icon is naturally wider than high, I propose to put it at 16px wide and set its height at 14px (natural height would be 14.3px). This will give it the same area as the files icon except in "landscape" mode instead of "protrait" if I can use these terms. Below is a preview of the result:

![image](https://user-images.githubusercontent.com/22901938/96942592-98294b80-14a3-11eb-9d4d-1f60e7bf6cae.png)

DrGFreeman replied:
Above proposal implemented in 3d28682.

rmol replied:
That looks better to me. @eloquence and @ninavizz?

rmol commented on securedrop/journalist_templates/admin.html:
```diff
@@ -3,7 +3,10 @@
 <form action="{{ url_for('admin.add_user') }}">
-  <button type="submit" id="add-user"><i class="fa fa-plus"></i>{{ gettext('ADD USER') }}</button>
+  <button type="submit" id="add-user">
+    <img src="{{ url_for('static', filename='icons/plus.png') }}" class="icon" width="13" height="15" alt="plus-icon">
```
This would be an exception to my request to translate alt tags. Where the icon is a visual repetition of adjacent text, we should use `alt=""`, per [WCAG recommendations for decorative images](https://www.w3.org/WAI/tutorials/images/decorative/).

DrGFreeman replied:
`alt=""` used where the alt text does not add information relative to what the adjacent text already provides.

rmol commented on securedrop/journalist_templates/col.html:
```diff
@@ -4,7 +4,7 @@
   <p class="breadcrumbs">
     <a href="/">{{ gettext('All Sources') }}</a>
-    <i class="fa fa-chevron-right"></i>
+    <img src="{{ url_for('static', filename='icons/chevron-right.png') }}" class="icon" width="9" height="14" alt=">">
```
And here, I think I would use `alt=""` on the separator icon rather than have a screenreader say something like "link all sources greater than contemptible bodybuilding" (or whatever the source designation were, and as I don't have a screenreader set up at the moment, this is just a guess at what it might sound like).

Really, the breadcrumbs should be done in a more [semantic and accessible way](https://www.w3.org/TR/wai-aria-practices/#breadcrumb), but that's definitely out of scope for this PR. :slightly_smiling_face:

DrGFreeman replied:
`alt=""` used.

rmol commented on securedrop/journalist_templates/edit_account.html:
```diff
@@ -4,7 +4,7 @@
   {# Only admins may edit usernames and admin status #}
   <h1>{{ gettext('Edit user "{user}"').format(user=user.username) }}</h1>
-  <p><a href="/admin">{{ gettext('Back to admin interface') }}</a></p>
+  <p><a href="/admin">« {{ gettext('Back to admin interface') }}</a></p>
```
This is great for consistency, but it's made me realize that all the places we use this as a navigational symbol probably look pretty odd in languages that use it as a quotation mark. Our links must look like they have more to say. :laughing: 

That's _another_ issue to file.

DrGFreeman replied:
I could remove the character on all admin pages. Let me know what you think.

rmol replied:
I've filed #5596 for that.

rmol commented on securedrop/sass/modules/_cols.sass:
```diff
@@ -59,8 +59,17 @@
           padding: 3px
           margin-right: 5px
-          font-size: 12px
+          font-size: 0.9em
```
Nice, relative units FTW.

rmol reviewed (changes requested):
Very nice work. I have one suggestion about the height of the messages icon, and a request that you review the new `alt` tags -- there are quite a few that could probably be left empty, as they're equivalent to adjacent text, and those that should remain should also be marked for translation, so they're accessible in all our supported languages.

Functionally, this is great; I ran through the test plan and everything looks good.

**Journalist homepage `/`**

Login to Journalist Interface using Tor Browser with security settings set to "Safest".
- [x] Verify "Filter by codename" placeholder text of input box has
      capital "F". (NOTE: This is of course not visible at "Safest"; I
      verified it at "Standard".)
- [x] Verify all icons display and are properly positioned.
- [x] Hover on Delete button and verify white trash icon is visible on hover.
- [x] Hover on source star and verify orange star is visible.
- [x] Click star on a source and verify the star icon turns to orange.
- [x] Verify nothing is abnormal with the page layout.

**Source submissions page `/col/<filesystem_id>`**

Login source interface in a new browser tab.
Submit one message.
Submit one file.

Refresh the JI browser tab.
Click on the new source's codename.
- [x] Verify the chevron icon displays between the All Source link and the source codename.
- [x] Verify all button icons display and are properly positioned.
- [x] Hover on Delete Selected button and verify the white trash icon is visible.
- [x] Verify nothing is abnormal with the page layout.
Download one source message.
Submit two replies.

Refresh Source Interface browser tab.
Delete one reply.

Refresh the JI browser tab.
- [x] Verify all icons display left of the source codename (open envelope, closed envelope, reply (unread), check (read reply) and are properly positioned.
- [x] Verify the rightmost icons display (files, messages, reply) and are properly positioned.
- [x] Scroll to bottom and verify trash icon displays on DELETE SOURCE AND SUBMISSIONS button.
- [x] Hover on DELETE SOURCE AND SUBMISSIONS button and verify that white trash icon is displayed.

**Admin page `/admin/`**

Click the Admin link.
- [x] Verify the plus and pencil icons display in the ADD USER and INSTANCE CONFIG buttons.
- [x] Verify the edit and trash icons display on each journalist row.
- [x] Verify the vertical spacing between the rows is similar.
- [x] Hover on the second journalist trash icon and verify the darker red trash icon displays.
- [x] Verify nothing is abnormal with the page layout

**Add User page `/admin/add`**

Click the ADD USER button.
- [x] Verify the plus icon displays in the ADD USER button.
- [x] Verify nothing is abnormal with the page layout.
Click the *Back to admin interface* link.
Additional
**Edit User page `/admin/edit/<journalist_id>`**

Click the edit icon on one journalist row.
- [x] Verify the "«" character displays at the begining of the *Back
      to admin interface* link.
- [x] Verify the "sync" icon displays on the RESET PASSWORD, RESET MOBILE APP CREDENTIALS and RESET SECURITY KEY CREDENTIALS buttons.
- [x] Verify nothing is abnormal with the page layout.
Click the *Back to admin interface* link.

**Instance Configuration page `/admin/config`**

Click the INSTANCE CONFIG button.
- [x] Verify the bell icon displays on the SEND TEST OSSEC ALERT button.
- [x] Verify the edit (pencil) icon displays on the UPDATE LOGO and UPDATE SUBMISSION PREFERENCES buttons.
- [x] Verify nothing is abnormal with the page layout.

**Flag source for reply success page `/flag`**

Shutdown the server and temporarily change line 250 of `securedrop/source_app/main.py` from `if entropy_avail >= 2400:` to `if entropy_avail >= 2400 and False:`.
Restart the server, log in as a new source and submit a message.
View this new source via the Journalist Interface.
Click FLAG THIS SOURCE FOR REPLY.
- [x] Verify the info circle icon is displays and is properly positioned.
- [x] Verify nothing is abnormal with the page layout.

DrGFreeman committed:
Update icons "alt" values

Set alt="" where the icon does not add information not already conveyed
by the adjacent text. Use gettext() for alt values of other icons to allow
for translation of alt text.

DrGFreeman committed:
Enlarge messages in source row icon

Enlarge the messages icon, while maintining its natural proportions, so that
it has the same footprint as the adjacent files icon (16x14px vs 14x16px
respectively).

DrGFreeman committed:
Update icons alt attribute in col.html

DrGFreeman commented:
@rmol, I either set `alt=""` or used `gettext()` for the `alt` attributes as suggested. To help with the review, below is a visual map of where I set `alt=""` (red) or used `gettext()` (green):

![image](https://user-images.githubusercontent.com/22901938/96945999-26ee9600-14ad-11eb-85cf-b57ec47199c7.png)

![image](https://user-images.githubusercontent.com/22901938/96946023-366ddf00-14ad-11eb-8339-d5eac9e20047.png)

![image](https://user-images.githubusercontent.com/22901938/96946041-4685be80-14ad-11eb-8efd-1f89224b2af8.png)

![image](https://user-images.githubusercontent.com/22901938/96946045-4be30900-14ad-11eb-80c6-9041922cbdcf.png)

![image](https://user-images.githubusercontent.com/22901938/96946057-530a1700-14ad-11eb-97b8-9cd195c858fc.png)

![image](https://user-images.githubusercontent.com/22901938/96946076-5b625200-14ad-11eb-9da6-15e5da5d4603.png)

![image](https://user-images.githubusercontent.com/22901938/96946086-64532380-14ad-11eb-92fe-2df552f7f193.png)

Let me know if this is OK or if you'd like changes.

DrGFreeman committed:
Add gettext to disabled trash icon alt attribute

DrGFreeman committed:
Manual HTML cleanup

Clean empty <i> tag and orphan </i> tag.

rmol commented:
> I either set `alt=""` or used `gettext()` for the `alt` attributes as suggested. To help with the review, below is a visual map of where I set `alt=""` (red) or used `gettext()` (green):
> [...]
> Let me know if this is OK or if you'd like changes.

OK? _OK_? :open_mouth: This is _fantastic_! Thank you so much for going to this trouble. I'm going to run through everything again with a screenreader, and this is going to make it so much easier to make sure I've tested all the changes.

rmol commented:
I'm approving this, but will wait to hear what others think of the message icon sizing before merging. 

Having gone through these pages with a screenreader, the new alt text works pretty well. Right now you hear things like "download icon two unread", or "unstarred icon push button", but we have much bigger usability problems here for screenreader users, and if we address those I think we're going to want to change the alt text again to work better in a different context. The star/unstar buttons, for example, with the right alt text and ARIA attributes could be made to say something like "source stellar bronchoscope, star toggle button, not pressed", which gives a clearer idea of the button's current and potential state. 

Thank you again, @DrGFreeman, for the great work and making review so much easier. I really appreciate it.

ninavizz commented:
@DrGFreeman These look phenomenal, THANK YOU!!

FWIW, some icons are taller than they are wide, and that's ok. The generalized footprint that icons occupy all taking-up the same volume of space, is generally more important. These look lovely. Merge away!
